### PR TITLE
Weighted losses for multitask

### PR DESCRIPTION
**Pull request by allaffa**
~~change back WIP to work on the unit_test for multitask~~
Ready for review @allaffa  @streeve  

There is a concern about the DDP implementation(in a different PR). Should the batch_size in DDP  be batch_size /= world_size? I am wondering if the repeated failures of "mpirun -n 2" only (where "mpirun -n 1" passed) are related to DDP implementation.

### CODE REVIEW

File: utils/visualizer.py [hunk 1]
```diff
@@ -49,16 +52,468 @@ def __convert_to_list(self):
...
-        plt.ylabel("Predicted value")
+        # self.__convert_to_list()
+        nshape = np.asarray(self.predicted_values).shape
+        if nshape[1] == 1:
+            fig = plt.figure(figsize=(15, 4.5))
+            plt.subplots_adjust(
+                left=0.08, bottom=0.15, right=0.95, top=0.925, wspace=0.35, hspace=0.1
+            )
+            ax = plt.subplot(1, 3, 1)
+            plt.scatter(self.true_values, self.predicted_values)
```
Comment (streeve) [inline utils/visualizer.py:87]: You could create a new function like `__scatter_impl(x, y,  title, xlabel, ylabel, xlim, ylim)`. Then a similar `__plot_impl(x, y,  title, xlabel, ylabel, xlim, ylim)` and reuse them throughout the class

**Review by streeve — CHANGES_REQUESTED** [review]
First round of review on this new PR

File: models/GATStack.py [hunk 2]
```diff
@@ -19,6 +19,9 @@ def __init__(
         dropout: float = 0.25,
         num_conv_layers: int = 16,
         num_shared: int = 1,
+        ihpwloss: int = 1,  # if =1, considering weighted losses for different tasks and treat the weights as hyper parameters
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/GATStack.py:22]: What do you think about changing the names as following?
ihpwloss --> ilossweights_hyperp
inllloss --> ilossweights_nll
hweights --> loss_weights

[the remark this answers is not in the document]
Comment (streeve) [reply models/GATStack.py:22]: I like those better, but I think we need documentation of what these are (there is a lot of good existing documentation, but we're also missing a lot). I'm still not sure what they are

File: models/Base.py [hunk 3]
```diff
@@ -26,19 +38,40 @@ def _multihead(self, output_dim: int, num_nodes: int, num_shared: int):
...
+                for ihead in range(self.num_heads)
             ]
 
         self.num_heads = len(outputs_dims)  # number of heads/tasks
+        self.head_dims = outputs_dims
         self.heads = ModuleList()
+        self.inllloss = inllloss
+        self.ihpwloss = ihpwloss
+        if self.ihpwloss * self.inllloss == 1:
+            sys.exit("Error: both ihpwloss and inllloss are set to 1.")
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/Base.py]: Fixed in 66c3241

File: models/Base.py [hunk 4]
```diff
@@ -67,11 +100,74 @@ def loss(self, pred, value):
...
+        return torch.sqrt(F.mse_loss(pred, value)), [], []
+
+    def loss_NLL(self, pred, value):  # negative log likelihood loss
+        # uncertainty to weigh losses in https://openaccess.thecvf.com/content_cvpr_2018/papers/Kendall_Multi-Task_Learning_Using_CVPR_2018_paper.pdf
+        pred_shape = pred.shape
+        value_shape = value.shape
+        if pred_shape[0] != value_shape[0]:
+            value = torch.reshape(value, (pred_shape[0], -1))
+            if value.shape[1] != pred.shape[1] - self.num_heads:
+                sys.exit(
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/Base.py]: 66c3241

File: models/Base.py [hunk 5]
```diff
@@ -26,19 +38,40 @@ def _multihead(self, output_dim: int, num_nodes: int, num_shared: int):
...
         self.num_heads = len(outputs_dims)  # number of heads/tasks
+        self.head_dims = outputs_dims
         self.heads = ModuleList()
+        self.inllloss = inllloss
+        self.ihpwloss = ihpwloss
+        if self.ihpwloss * self.inllloss == 1:
+            sys.exit("Error: both ihpwloss and inllloss are set to 1.")
+        if self.ihpwloss == 1:
+            if len(hweights) != self.num_heads:
+                sys.exit(
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/Base.py]: 66c3241

File: models/GATStack.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/GATStack.py]: Updated in 66c3241

File: models/PNNStack.py [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import torch
 import torch.nn.functional as F
 from torch.nn import ModuleList
+from torch.nn import Sequential, ReLU, Linear
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/PNNStack.py]: Removed in 66c3241

File: utils/visualizer.py (hunk 1 above)
Comment (pzhanggit) [reply utils/visualizer.py]: Updated in 66c3241

File: models/Base.py [hunk 7]
```diff
@@ -67,11 +100,74 @@ def loss(self, pred, value):
...
+        elif self.ihpwloss == 1:
+            return self.loss_hpweighted(pred, value)
+
+        pred_shape = pred.shape
+        value_shape = value.shape
+        if pred_shape != value_shape:
+            value = torch.reshape(value, pred_shape)
+        return torch.sqrt(F.mse_loss(pred, value)), [], []
+
+    def loss_NLL(self, pred, value):  # negative log likelihood loss
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/Base.py]: updated

File: utils/utils.py [hunk 8]
```diff
@@ -17,6 +17,8 @@
     Dataset,
 )
 from utils.visualizer import Visualizer
+import numpy as np
+import matplotlib.pyplot as plt
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply utils/utils.py]: removed

**Comment by pzhanggit** [comment]
This PR is waiting for #33

**Comment by pzhanggit** [comment]
@allaffa  Max, could you also take a look at the unit test part if the changes make sense? Thanks.

File: data_utils/dataset_descriptors.py [hunk 9]
```diff
@@ -26,9 +26,9 @@ class Dataset(Enum):
 
     FePt = "FePt_32atoms"
     CuAu = "CuAu_32atoms"
-    CuAu_FePt_SHUFFLE = "CuAu_FePt_SHUFFLE"
-    CuAu_TRAIN_FePt_TEST = "CuAu_TRAIN_FePt_TEST"
-    FePt_TRAIN_CuAu_TEST = "FePt_TRAIN_CuAu_TEST"
+    # CuAu_FePt_SHUFFLE = "CuAu_FePt_SHUFFLE"
```
Comment (allaffa) [inline data_utils/dataset_descriptors.py:29]: I would delete these comments.

File: test_trained_model.py [hunk 10]
```diff
@@ -26,66 +26,76 @@ def test_trained_model(config_file: str = None, chosen_model: torch.nn.Module =
...
-    }
-    chosen_dataset_option = None
-    for dataset in dataset_options.values():
-        if dataset.value == config["dataset_option"]:
-            chosen_dataset_option = dataset
 
     train_loader, val_loader, test_loader = dataset_loading_and_splitting(
         config=config,
-        chosen_dataset_option=chosen_dataset_option,
+        chosen_dataset_option=config["Dataset"]["name"],
```
Comment (allaffa) [inline test_trained_model.py:32]: I thought that test_trained_model would disappear. 
@streeve @pzhanggit  why is it still here?

Comment (streeve) [reply test_trained_model.py:32]: You call it inside `test_random_graphs`

File: utils/utils.py [hunk 11]
```diff
@@ -71,55 +72,188 @@ def train_validate_test_normal(
...
+    vallib = []
+    testlib = []  # total loss tracking for train/vali/test
+    tasklib = []
+    tasklib_test = []
+    tasklib_vali = []  # loss tracking for summation across all atoms/nodes
+    tasklib_nodes = []
+    tasklib_test_nodes = []
+    tasklib_vali_nodes = []  # probably not needed
+
+    if isinstance(model_wrapper, torch.nn.parallel.distributed.DistributedDataParallel):
```
Comment (streeve) [inline utils/utils.py:90]: This could be a separate PR

File: utils/utils.py [hunk 12]
```diff
@@ -71,55 +72,188 @@ def train_validate_test_normal(
     scheduler,
     config,
     model_with_config_name,
+    plot_init_sol=True,
```
Comment (streeve) [inline utils/utils.py:75]: What is sol?

**Review by streeve — COMMENTED** [review]
This is becoming a large PR - weighted loss + visualization + new input file. I get that they're all connected, but as much as we can avoid this we should

We need to be careful about letting anything else creep into this

File: test_trained_model.py (hunk 10 above)
Comment (pzhanggit) [reply test_trained_model.py:32]: Are there any reasons that make us want it disappear? The current implementation looks good to me. 
The are some details we can improve after moving to ORNL.

File: data_utils/raw_dataset_loader.py [hunk 13]
```diff
@@ -68,11 +94,18 @@ def __transform_input_to_data_object(self, lines: [str]):
         Data
             Data object representing structure of an atom.
         """
+
+        # fixme, not flexible, improve later
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py:98]: No, I mean the raw file format that this function can handle is not flexible enough. It works for the LSMS format so far. This part is purely for data formatting.

File: utils/utils.py (hunk 12 above)
Comment (pzhanggit) [reply utils/utils.py:75]: sol=solution. I will change it.

File: utils/utils.py (hunk 11 above)
Comment (pzhanggit) [reply utils/utils.py:90]: What could be?

File: data_utils/raw_dataset_loader.py [hunk 14]
```diff
@@ -30,17 +30,35 @@ def load_raw_data(self, dataset_path: str):
         dataset_path: str
             Directory path where raw files are stored.
         """
+
+        atom_feature_dim = config["atom_featurs"]["dim"]
+        atom_feature_loc = config["atom_featurs"]["locations"]
+        prop_dim = config["properties"]["dim"]
+        prop_loc = config["properties"]["locations"]
+        LSMS = True
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py:38]: I did not use "LSMS" because there are some specific operations, e.g., line 130, associated with the material datasets we have. Not sure if that exists in all LSMS.

File: utils/utils.py (hunk 11 above)
Comment (streeve) [reply utils/utils.py:90]: A `model_wrapper` function that returns model/module that we can call everywhere

File: utils/utils.py [hunk 15]
```diff
@@ -71,55 +72,188 @@ def train_validate_test_normal(
...
+        tasklib.append(train_taskserr)
+        tasklib_vali.append(val_taskserr)
+        tasklib_test.append(test_rmse[1])
+
+        tasklib_nodes.append(train_taskserr_nodes)
+        tasklib_vali_nodes.append(val_taskserr_nodes)
+        tasklib_test_nodes.append(test_rmse[2])
+
+        ###tracking the solution evolving with training
+        if plot_hist_sol:
```
Comment (allaffa) [inline utils/utils.py:155]: the plot of the solution should be a separate function. The visualization techniques should be part of the training. 
In fact, I would be nice if the training function returned the history of training-valdiation-test-losses, and then these were used by another function

File: utils/utils.py [hunk 16]
```diff
@@ -71,55 +72,188 @@ def train_validate_test_normal(
...
-            for iout in range(len(predicted_values[0])):
-                predicted_values[isamp][iout] = (
-                    predicted_values[isamp][iout]
-                    * (y_minmax[iout][1] - y_minmax[iout][0])
-                    + y_minmax[iout][0]
+        true_values, predicted_values = output_denormalize(
+            config["Target_dataset"]["y_minmax"], true_values, predicted_values
+        )
+
+    for ihead in range(model.num_heads):
```
Comment (allaffa) [inline utils/utils.py:194]: the plot of the solution should be a separate function. The visualization techniques should be part of the training.
In fact, I would be nice if the training function returned the history of training-valdiation-test-losses, and then these were used by another function

File: data_utils/raw_dataset_loader.py [hunk 17]
```diff
@@ -84,19 +117,33 @@ def __transform_input_to_data_object(self, lines: [str]):
...
+                num_of_protons = node_feature[0]
+                charge_density = node_feature[1]
+                magnetic_moment = float(node_feat[6].strip())
+                charge_density = charge_density - num_of_protons
+                node_feature[1] = charge_density
+                if node_feature_dim[2] == 3:
+                    magnetic_moment_x = float(node_feat[7].strip())
+                    magnetic_moment_y = float(node_feat[8].strip())
+                    magnetic_moment_z = float(node_feat[9].strip())
+                    magnetic_moment_x *= magnetic_moment
```
Comment (allaffa) [inline data_utils/raw_dataset_loader.py:136]: I know that you disagree on adding comments, but explaining why you do 
magnetic_moment_x *= magnetic_moment
would be nice. 
Put yourself in the shoes of a person that is looking at the code for the first time. 
When we talked about this, you said "A person just has to read the code and he/she understands". This code will eventually explore in a lot of capabilities one day. Do you really believe that?

Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py:136]: You are reviewing an outdated version for this part. It has been updated in the general_improvement branch that has been reviewed and merged by you into the main branch.

Comment (allaffa) [reply data_utils/raw_dataset_loader.py:136]: LOL

File: utils/utils.py (hunk 11 above)
Comment (pzhanggit) [reply utils/utils.py:149]: Yeah, I like the idea. Very neat. Let's do this in a separate PR.

**Review by streeve — CHANGES_REQUESTED** [review]
Let's wait on #55 to merge. Can you also post some example visualization?

File: data_utils/raw_dataset_loader.py [hunk 18]
```diff
@@ -128,12 +128,10 @@ def __transform_input_to_data_object_base(
                 for icomp in range(node_feature_dim[item]):
                     it_comp = node_feature_col[item] + icomp
                     node_feature.append(float(node_feat[it_comp].strip()))
-
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply data_utils/raw_dataset_loader.py:131]: I like the current version better. Consistent coding format. Let's keep it this way?

File: models/Base.py [hunk 19]
```diff
@@ -2,15 +2,26 @@
 from torch.nn import ModuleList, Sequential, ReLU, Linear
 import torch.nn.functional as F
 from torch_geometric.nn import global_mean_pool
+from torch.nn import GaussianNLLLoss
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/Base.py:5]: Yes, it requires 1.8+, according to https://discuss.pytorch.org/t/attributeerror-module-torch-nn-has-no-attribute-gaussiannllloss/121425

**Comment by pzhanggit** [comment]
> Let's wait on #55 to merge. Can you also post some example visualization?

Something like plot user guide? Where to post?

File: data_utils/raw_dataset_loader.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (streeve) [reply data_utils/raw_dataset_loader.py:131]: No problem - I thought it was from a rebase

File: models/Base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (streeve) [reply models/Base.py:5]: Ok, I added a commit for that - good news is that the cache reset like it should

**Review by streeve — COMMENTED** [review]
> Something like plot user guide? Where to post?

I was thinking example png files from the new visualization code here in the pull request (you can attach to a comment)

**Comment by pzhanggit** [comment]
> > Something like plot user guide? Where to post?
> 
> I was thinking example png files from the new visualization code here in the pull request (you can attach to a comment)

Got you. See updates in commit c60a22.

**Comment by pzhanggit** [comment]
Waiting for PR#66

**Comment by pzhanggit** [comment]
Not ready for review yet. Need to figure out:
1. how to add test_graphs_multihead.py into CI test
2. Improve GAT and GIN for node prediction

File: utils/visualizer.py [hunk 20]
```diff
@@ -1,62 +1,635 @@
...
+
+    def __err_condmean(self, data1, data2, weight=1.0):
+        errabs = np.abs(np.hstack(data1) - np.hstack(data2)) * weight
+        hist2d_pasr, xedge_pasr, yedge_pasr = np.histogram2d(
+            np.hstack(data1), errabs, bins=50
+        )
+        xcen_pasr = 0.5 * (xedge_pasr[0:-1] + xedge_pasr[1:])
+        ycen_pasr = 0.5 * (yedge_pasr[0:-1] + yedge_pasr[1:])
+        hist2d_pasr = hist2d_pasr / np.amax(hist2d_pasr)
+        mean1d_cond = np.dot(hist2d_pasr, ycen_pasr) / np.sum(hist2d_pasr, axis=1)
```
Comment (streeve) [inline utils/visualizer.py:78]: `RuntimeWarning: invalid value encountered in true_divide`

**Review by streeve — COMMENTED** [review]
More small things. Tests are looking better!

File: utils/utils.py [hunk 21]
```diff
@@ -196,67 +291,114 @@ def train(loader, model, opt, output_dim):
...
+    return (
+        total_error / len(loader.dataset),
+        tasks_error / len(loader.dataset),
+        tasks_noderr / len(loader.dataset),
+    )
 
 
 @torch.no_grad()
-def test(loader, model, output_dim):
+def test(loader, model_wrapper):
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply utils/utils.py]: Updated in e312a340

File: utils/utils.py [hunk 22]
```diff
@@ -127,40 +128,130 @@ def train_validate_test_normal(
     scheduler,
     config,
     model_with_config_name,
+    plot_init_solution=True,
+    plot_hist_solution=False,
 ):
 
     num_epoch = config["Training"]["num_epoch"]
+    # total loss tracking for train/vali/test
+    trainlib = []
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply utils/utils.py]: Renamed in e312a340

File: utils/visualizer.py (hunk 20 above)
Comment (pzhanggit) [reply utils/visualizer.py]: It is caused by divide by zero I think. Added a small number in e312a340. It should be gone now.
